Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -536,43 +536,91 @@ u16 GCMemcard::DEntry_BlockCount(u8 index) const
   return blocks;
 }
 
-u32 GCMemcard::DEntry_CommentsAddress(u8 index) const
+std::optional<std::vector<u8>> GCMemcard::GetSaveDataBytes(u8 save_index, size_t offset,
+                                                           size_t length) const
 {
-  if (!m_valid || index >= DIRLEN)
-    return 0xFFFF;
+  if (!m_valid || save_index >= DIRLEN)
+    return std::nullopt;
 
-  return GetActiveDirectory().m_dir_entries[index].m_comments_address;
-}
+  const DEntry& entry = GetActiveDirectory().m_dir_entries[save_index];
+  const BlockAlloc& bat = GetActiveBat();
+  const u16 block_count = entry.m_block_count;
+  const u16 first_block = entry.m_first_block;
+  const u32 block_max = BAT_SIZE + MC_FST_BLOCKS;
+  if (block_count == 0xFFFF || first_block < MC_FST_BLOCKS || first_block >= block_max)
+    return std::nullopt;
 
-std::string GCMemcard::GetSaveComment1(u8 index) const
-{
-  if (!m_valid || index >= DIRLEN)
-    return "";
+  const u32 file_size = block_count * BLOCK_SIZE;
+  if (offset >= file_size)
+    return std::nullopt;
 
-  u32 Comment1 = GetActiveDirectory().m_dir_entries[index].m_comments_address;
-  u32 DataBlock = GetActiveDirectory().m_dir_entries[index].m_first_block - MC_FST_BLOCKS;
-  if ((DataBlock > m_size_blocks) || (Comment1 == 0xFFFFFFFF))
+  const size_t bytes_to_copy = std::min(length, file_size - offset);
+  std::vector<u8> result;
+  result.reserve(bytes_to_copy);
+
+  u16 current_block = first_block;
+  size_t offset_in_current_block = offset;
+  size_t bytes_remaining = bytes_to_copy;
+
+  // skip unnecessary blocks at start
+  while (offset_in_current_block >= BLOCK_SIZE)
   {
-    return "";
+    offset_in_current_block -= BLOCK_SIZE;
+    current_block = bat.GetNextBlock(current_block);
+    if (current_block < MC_FST_BLOCKS || current_block >= block_max)
+      return std::nullopt;
+  }
+
+  // then copy one block at a time into the result vector
+  while (true)
+  {
+    const GCMBlock& block = m_data_blocks[current_block - MC_FST_BLOCKS];
+    const size_t bytes_in_current_block_left =
+        std::min(BLOCK_SIZE - offset_in_current_block, static_cast<size_t>(BLOCK_SIZE));
+    const size_t bytes_in_current_block_left_to_copy =
+        std::min(bytes_remaining, bytes_in_current_block_left);
+
+    for (u32 i = 0; i < bytes_in_current_block_left_to_copy; ++i)
+      result.push_back(block.m_block[offset_in_current_block + i]);
+
+    bytes_remaining -= bytes_in_current_block_left_to_copy;
+    if (bytes_remaining == 0)
+      break;
+
+    offset_in_current_block = 0;
+    current_block = bat.GetNextBlock(current_block);
+    if (current_block < MC_FST_BLOCKS || current_block >= block_max)
+      return std::nullopt;
   }
-  return std::string((const char*)m_data_blocks[DataBlock].m_block.data() + Comment1,
-                     DENTRY_STRLEN);
+
+  return result;
```

Comment:
This should be std::moved, otherwise it'll get copied, given the return type isn't `std::vector<u8>`. Another way to make this more explicit is:

```cpp
return {std::move(result)};
```

or

```cpp
return std::make_optional(std::move(result));
```